sway/input/seatop_default.c

```diff
@@ -747,6 +750,208 @@ static void handle_pointer_axis(struct sway_seat *seat,
 	}
 }
 
+static uint32_t wl_swipe_button_base(uint8_t fingers) {
+	if (fingers == 3) {
+		return SWAY_SWIPE_3;
+	} else if (fingers == 4) {
+		return SWAY_SWIPE_4;
+	}
+
+	return 0;
+}
+
+static void handle_swipe_begin(struct sway_seat *seat,
+		struct wlr_event_pointer_swipe_begin *event) {
+	struct sway_input_device *input_device =
+		event->device ? event->device->data : NULL;
+	struct sway_cursor *cursor = seat->cursor;
+	struct seatop_default_event *e = seat->seatop_data;
+
+	// Determine what's under the cursor
+	struct wlr_surface *surface = NULL;
+	double sx, sy;
+	struct sway_node *node = node_at_coords(seat,
+			cursor->cursor->x, cursor->cursor->y, &surface, &sx, &sy);
+	struct sway_container *cont = node && node->type == N_CONTAINER ?
+		node->sway_container : NULL;
+	enum wlr_edges edge = cont ? find_edge(cont, surface, cursor) : WLR_EDGE_NONE;
+	bool on_border = edge != WLR_EDGE_NONE;
+	bool on_titlebar = cont && !on_border && !surface;
+	bool on_contents = cont && !on_border && surface;
+	bool on_workspace = node && node->type == N_WORKSPACE;
+
+	bool handled = false;
+
+	// Gather information needed for mouse bindings
+	struct wlr_keyboard *keyboard = wlr_seat_get_keyboard(seat->wlr_seat);
+	uint32_t modifiers = keyboard ? wlr_keyboard_get_modifiers(keyboard) : 0;
+	struct wlr_input_device *device =
+		input_device ? input_device->wlr_device : NULL;
+	char *dev_id = device ? input_device_get_identifier(device) : strdup("*");
+
+	// zero for both turns below loop into a no-op
+	uint32_t button_base = wl_swipe_button_base(event->fingers);
+	uint32_t button_end = 0;
+	if (button_base != 0) {
+		button_end = button_base + SWAY_SWIPE_DIR_COUNT;
+	}
+
+	/**
+	 * Handle mouse bindings - if there's any swipe bindings for this
+	 * context. Here we must swallow all swipes if there's even one
+	 * possible binding because we have no way of knowing yet what kind of
+	 * swipe this is going to be. It's in the nature of the thing because
+	 * the user may turn a left into an upwards swipe mid-way.
+	 */
```

Comment:
At least Safari and Firefox seem to do two- and three-finger gestures ever since 2009: http://www.macyourself.com/2009/09/18/mac-browser-showdown-multi-touch-gestures-in-safari-vs-firefox/. I think to have played around with that in Firefox at one point. The mail app (which I've never used) seems to have (at least at one point had) a swipe to archive function: https://osxdaily.com/2015/12/02/change-mail-swipe-gesture-mac-os-x/.

Reply:
So if you use a swipe over the mail app it will archive something instead of switch workspaces? That's counter to how sway normally does bindings, but if we wanted we could coerce wlroots into telling us which apps care about swipes and give them first chance. It might be fine if only a minority of clients really care about swipes. Though if some apps got priority on swipes it might be better to just have a command the user can use with `for_window` to choose which ones...

I guess the current behavior is fine, I'm just not used to swipes.

Reply:
I'd love to have some gestures in Firefox for navigation but was not able to get it on Linux working yet. This discussion has sparked my interest again. On the other hand I do not want to drag this out. So I'd suggest merging as-is and perhaps follow up with a separate PR regarding pass-through to swipe-enabled clients. Right now this would leave the user with a binary decision for each type of swipe: Either have a compositor-level binding or pass-through to clients. What d'you think?